### DISCUSSION TITLE
What's the parameter / component which caused this error ? ---> [backend/drm/drm.c:60] PRIME import not supported

### DISCUSSION THREAD

**Discussion by Marietto2008** [WayfireWM/wayfire#2198] [Q&A]
I have two FreeBSD 14 installations. I state that my PC has 3 gpus.

1) Intel Corporation CoffeeLake-S GT2 [UHD Graphics 630]
2) Nvidia GTX 1060
3) Nvidia RTX 2080ti

On the first FreeBSD 14 system I've chosen as the main GPU the Geforce GTX 1060 and the RTX 2080 ti is reserved for the Linux virtual machines where I make the passthru ; the Intel gpu is not even shown on the system. With this setup,Wayland / Wayfire works. On the secondary installation I have chosen the Intel Corporation CoffeeLake-S GT2 [UHD Graphics 630] as main GPU ; the GTX 1060 is secondary and the RTX 2080 it is reserved for the bhyve / Linux vms. Even here it works.

Now,probably I have configured wayland / wayfire in a different way on the secondary FreeBSD 14 installation,because although at the beginning I used the same GPU setup of the system 1 (main GPU Geforce GTX 1060 ; RTX 2080 ti reserved for Linux and Intel gpu not shown by the system),at some point it stopped working. I don't know what I have changed. The fact is that I got this error message :

```
[src/main.cpp:334] Starting wayfire version 0.9.0
[libseat] [libseat/libseat.c:66] Seat opened with backend 'consolekit2'
[libseat] [libseat/backend/consolekit2.c:92] Ping failed: org.freedesktop.DBus.Error.AccessDenied: Rejected send message, 1 matched rules; type="method_call", sender=":1.9" (uid=1001 pid=4931 comm="wayfire") interface="org.freedesktop.DBus.Peer" member="Ping" error name="(unset)" requested_reply="0" destination="org.freedesktop.ConsoleKit" (uid=0 pid=4895 comm="/usr/local/sbin/console-kit-daemon --no-daemon")
[backend/session/session.c:109] Successfully loaded libseat session
[backend/backend.c:213] Found 1 GPUs
[backend/drm/backend.c:202] Initializing DRM backend for /dev/dri/card0 (nvidia-drm)
[backend/drm/drm.c:60] PRIME import not supported
[backend/backend.c:221] Failed to create DRM backend
[backend/backend.c:232] Could not successfully create backend on any GPU
[backend/backend.c:376] Failed to open any DRM device
[src/main.cpp:131] Fatal error: Segmentation fault
#1  0x36b365 <_ZL14signal_handleri+0xb5> at /opt/wayfire/bin/wayfire
#2  0x83861753f <pthread_sigmask+0x53f> at /lib/libthr.so.3
#3  0x838616afb <pthread_setschedparam+0x83b> at /lib/libthr.so.3
#4  0x8207ab2d3 <__gxx_personality_v0+0x8203788d3> at ???
#5  0x821c361d4 <wlr_backend_get_drm_fd+0x4> at /opt/wayfire/lib//libwlroots.so.12
#6  0x36a443 <main+0x423> at /opt/wayfire/bin/wayfire
#7  0x82be97afa <__libc_start1+0x12a> at /lib/libc.so.7
```

and I had to change GPU setup. I had to chosen the Intel Corporation CoffeeLake-S GT2 [UHD Graphics 630] as main GPU ; the GTX 1060 as secondary and the RTX 2080 reserved for Linux. I'm really interested to understand what's the parameters which forced me to change the layout of the GPU.

Based on the error :

`0x821c361d4 <wlr_backend_get_drm_fd+0x4> at /opt/wayfire/lib//libwlroots.so.12`

it seems that it could be "wlroots". Maybe using a new version of wlroots as I did,something is changed and it started to refuse to work on the gpu setup that I've used on the system 1.
Point is that for me is more comfortable to keep the nvidia gpu 1060 as primary and the rtx 2080 ti reserved and I don't need the intel integrated gpu.

**Comment by Marietto2008** [comment]
@ammen99 ; @soreau

**Comment by soreau** [comment]
> [backend/drm/drm.c:60] PRIME import not supported

This points to wlroots [here](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/backend/drm/drm.c?ref_type=heads#L60). You can see the `DRM_CAP_PRIME` value with a tool such as [drm_info](https://gitlab.freedesktop.org/emersion/drm_info) and compare with the value of [DRM_PRIME_CAP_IMPORT](https://gitlab.freedesktop.org/mesa/drm/-/blob/main/include/drm/drm.h?ref_type=heads#L685). Basically, if `DRM_CAP_PRIME` is an even number, there is no support for `DRM_PRIME_CAP_IMPORT`, so the error is thrown.

**Comment by soreau** [comment]
According to [the DRM_PRIME_CAP_IMPORT documentation](https://dri.freedesktop.org/docs/drm/gpu/drm-uapi.html#c.DRM_PRIME_CAP_IMPORT):

"Starting from kernel version 6.6, this bit is always set in [DRM_CAP_PRIME](https://dri.freedesktop.org/docs/drm/gpu/drm-uapi.html#c.DRM_CAP_PRIME)."

So I would guess your kernel is not up to date or the bits haven't been ported to your OS kernel yet.

**Reply by Marietto2008** [reply]
I've installed the tool called "drm_info" and I'm going to run it on both the two installations of FreeBSD 14 that I have,to see the differences,with the goal to understand what changes from one to another. You say that the kernel is not up to date. But this could be valid for Linux,but I'm using FreeBSD. And I'm using the same FreeBSD version and kernel for both the installations. So the reason could be different : my guess is that I've installed a version of wlroots that's too new and maybe not compatible with some component of FreeBSD.

**Reply by Marietto2008** [reply]
I solved the mistery. On the system where the error is "PRIME import not supported" I have installed the nvidia drivers from packages ; in this scenario I'm forced to keep the Intel GPU as primary if I want to run Wayland / Wayfire. On the system where I have installed the nvidia drivers from ports wayland / wayfire will work if I choose to have the GPU nvidia 1060 as primary. So,don't install the nvidia drivers from packages. (in both the systems I have installed the nvidia drivers vers. 550). I'm pretty sure of this,because both systems have the nvidia-drm.ko file of different size. So,choosing to install the nvidia drivers (or only the nvidia-drm-kmod file ?) from packages or from ports make the difference. Honestly I don't remember if when I have installed the nvidia-drivers from ports I have chosen or not to enable the PRIME option (if it is there).
